src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,877 @@
+          closest = k_behind;
+        }
+        // Get the minimum of the cell that's closest to an obstacle and the
+        // current distance to an obstacle, and multiply by the true sign of
+        // curVal
+        if (closest == INT_MAX)
+          closest--;
+        curVal = ((curVal > 0) - (curVal < 0)) *
+                 std::min(abs(std::max(curVal, -2147483647)), closest + 1);
+        sdfGrid[i][j][k] = curVal;
+      }
+    }
+  }
+  // second sweep
+  for (int i = m_voxelGridDimensions[0] - 1; i >= 0; i--) {
+    for (int j = m_voxelGridDimensions[1] - 1; j >= 0; j--) {
+      for (int k = m_voxelGridDimensions[2] - 1; k >= 0; k--) {
+        int curVal = getVoxel<int>(Mn::Vector3i(i, j, k), gridName);
+        if (curVal == 0)
+          continue;
+        int i_ahead = INT_MAX, j_ahead = INT_MAX, k_ahead = INT_MAX;
+        if (isValidIndex(Mn::Vector3i(i + 1, j, k))) {
+          i_ahead = abs(std::max(sdfGrid[i + 1][j][k], -2147483646));
+        }
+        if (isValidIndex(Mn::Vector3i(i, j + 1, k))) {
+          j_ahead = abs(std::max(sdfGrid[i][j + 1][k], -2147483646));
+        }
+        if (isValidIndex(Mn::Vector3i(i, j, k + 1))) {
+          k_ahead = abs(std::max(sdfGrid[i][j][k + 1], -2147483646));
+        }
+
+        int closest = INT_MAX - 1;
+        if (i_ahead <= j_ahead && i_ahead <= k_ahead) {
+          // i_ahead is closest to nearest obstacle.
+          closest = i_ahead;
+        } else if (j_ahead <= i_ahead && j_ahead <= k_ahead) {
+          // j_ahead is closest to nearest obstacle.
+          closest = j_ahead;
+        } else {
+          // k_ahead is closest or tied for closest to nearest obstacle.
+          closest = k_ahead;
+        }
+        // Get the minimum of the cell that's closest to an obstacle and the
+        // current distance to an obstacle, and multiply by the true sign of
+        // curVal
+        if (closest == INT_MAX)
+          closest--;
+        curVal = ((curVal > 0) - (curVal < 0)) *
+                 std::min(abs(std::max(curVal, -2147483647)), closest + 1);
+        sdfGrid[i][j][k] = curVal;
+      }
+    }
+  }
+}
+
+void VoxelGrid::generateEuclideanDistanceSDF(const std::string& gridName) {
+  // check to see if Interior/Exterior grid exists, if not, generate it
+  if (grids_.find("InteriorExterior") == grids_.end()) {
+    generateInteriorExteriorVoxelGrid();
+  }
+  // create new vector3Grid and fill data from interior/exterior grid
+  addGrid<Mn::Vector3>("ClosestBoundaryCell");
+  auto intExtGrid = getGrid<int>("InteriorExterior");
+  auto closestCellGrid = getGrid<Mn::Vector3>("ClosestBoundaryCell");
+
+  Mn::Debug() << m_voxelGridDimensions;
+  for (int i = 0; i < m_voxelGridDimensions[0]; i++) {
+    for (int j = 0; j < m_voxelGridDimensions[1]; j++) {
+      for (int k = 0; k < m_voxelGridDimensions[2]; k++) {
+        Mn::Vector3i index = Mn::Vector3i(i, j, k);
+        int label = getVoxel<int>(index, "InteriorExterior");
```

Comment:
And this is why I was suggesting to just remove the `getVoxel()` / `setVoxel()` APIs altogether -- to avoid using them by accident. (Yes, I know they had their use before you introduced strided array views, but now they don't anymore.) Because this means you do a redundant string lookup for *every voxel*, while you could do just *one*, above when you called the `getGrid()`, and then you can index the grid directly:

```suggestion
        int label = intExtGrid[i][j][k];
```

Same in all other cases where `getVoxel()` / `setVoxel()` is used -- fetch the grid once, index the view instead :)